File: frontend/src/types/wiki.ts
/*
```cypher
CREATE
  (f:File {name: "wiki.ts", type: "file", language: "typescript"}),
  (m:Module {name: "@/types/wiki", type: "module"}),
  (c1:Class {name: "WikiSpace", type: "class", language: "typescript", signature: "interface WikiSpace"}),
  (c2:Class {name: "WikiPage", type: "class", language: "typescript", signature: "interface WikiPage"}),
  (c3:Class {name: "TextPatch", type: "class", language: "typescript", signature: "interface TextPatch"}),
  (c4:Class {name: "WikiPageUpdate", type: "class", language: "typescript", signature: "interface WikiPageUpdate"}),
  (c5:Class {name: "WikiPresence", type: "class", language: "typescript", signature: "interface WikiPresence"}),
  (c6:Class {name: "WikiSyncSnapshot", type: "class", language: "typescript", signature: "interface WikiSyncSnapshot"}),
  (c7:Class {name: "CreateWikiSpaceRequest", type: "class", language: "typescript", signature: "interface CreateWikiSpaceRequest"}),
  (c8:Class {name: "CreateWikiPageRequest", type: "class", language: "typescript", signature: "interface CreateWikiPageRequest"}),
  (f)-[:CONTAINS]->(m),
  (m)-[:CONTAINS]->(c1),
  (m)-[:CONTAINS]->(c2),
  (m)-[:CONTAINS]->(c3),
  (m)-[:CONTAINS]->(c4),
  (m)-[:CONTAINS]->(c5),
  (m)-[:CONTAINS]->(c6),
  (m)-[:CONTAINS]->(c7),
  (m)-[:CONTAINS]->(c8);
```
*/

export interface WikiSpace {
  id: string;
  workspace_id: string;
  name: string;
  slug: string;
  description?: string | null;
  created_by_name: string;
  created_at: string;
  updated_at: string;
}

export interface WikiPage {
  id: string;
  workspace_id: string;
  space_id: string;
  parent_id?: string | null;
  title: string;
  slug: string;
  content_markdown: string;
  version: number;
  updated_by_name: string;
  created_at: string;
  updated_at: string;
}

export interface TextPatch {
  from: number;
  to: number;
  insert: string;
}

export interface WikiPageUpdate {
  id: string;
  page_id: string;
  client_id: string;
  base_version: number;
  version: number;
  patch: TextPatch;
  content_markdown: string;
  created_by_name: string;
  created_at: string;
}

export interface WikiPresence {
  page_id: string;
  client_id: string;
  user_id?: string | null;
  display_name: string;
  cursor_anchor?: number | null;
  cursor_head?: number | null;
  last_seen_at: string;
}

export interface WikiSyncSnapshot {
  page: WikiPage;
  updates: WikiPageUpdate[];
  collaborators: WikiPresence[];
}

export interface CreateWikiSpaceRequest {
  workspace_id: string;
  name: string;
  slug: string;
  description?: string;
}

export interface CreateWikiPageRequest {
  workspace_id: string;
  space_id: string;
  title: string;
  slug: string;
  content_markdown?: string;
}

File: backend/src/models/wiki.rs
/*
```cypher
CREATE
  (f:File {name: "wiki.rs", type: "file", language: "rust"}),
  (m:Module {name: "crate::models::wiki", type: "module"}),
  (c1:Class {name: "WikiSpace", type: "class", language: "rust", signature: "struct WikiSpace"}),
  (c2:Class {name: "WikiPage", type: "class", language: "rust", signature: "struct WikiPage"}),
  (c3:Class {name: "WikiPageUpdate", type: "class", language: "rust", signature: "struct WikiPageUpdate"}),
  (c4:Class {name: "WikiPresence", type: "class", language: "rust", signature: "struct WikiPresence"}),
  (c5:Class {name: "TextPatch", type: "class", language: "rust", signature: "struct TextPatch"}),
  (c6:Class {name: "CreateWikiSpaceRequest", type: "class", language: "rust", signature: "struct CreateWikiSpaceRequest"}),
  (c7:Class {name: "CreateWikiPageRequest", type: "class", language: "rust", signature: "struct CreateWikiPageRequest"}),
  (c8:Class {name: "ApplyWikiUpdateRequest", type: "class", language: "rust", signature: "struct ApplyWikiUpdateRequest"}),
  (c9:Class {name: "WikiPresenceRequest", type: "class", language: "rust", signature: "struct WikiPresenceRequest"}),
  (c10:Class {name: "WikiSyncQuery", type: "class", language: "rust", signature: "struct WikiSyncQuery"}),
  (c11:Class {name: "WikiSyncSnapshot", type: "class", language: "rust", signature: "struct WikiSyncSnapshot"}),
  (f)-[:CONTAINS]->(m),
  (m)-[:CONTAINS]->(c1),
  (m)-[:CONTAINS]->(c2),
  (m)-[:CONTAINS]->(c3),
  (m)-[:CONTAINS]->(c4),
  (m)-[:CONTAINS]->(c5),
  (m)-[:CONTAINS]->(c6),
  (m)-[:CONTAINS]->(c7),
  (m)-[:CONTAINS]->(c8),
  (m)-[:CONTAINS]->(c9),
  (m)-[:CONTAINS]->(c10),
  (m)-[:CONTAINS]->(c11);
```
*/

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, FromRow)]
pub struct WikiSpace {
    pub id: Uuid,
    pub workspace_id: Uuid,
    pub name: String,
    pub slug: String,
    pub description: Option<String>,
    pub created_by: Option<Uuid>,
    pub created_by_name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, FromRow)]
pub struct WikiPage {
    pub id: Uuid,
    pub workspace_id: Uuid,
    pub space_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub title: String,
    pub slug: String,
    pub content_markdown: String,
    pub version: i64,
    pub created_by: Option<Uuid>,
    pub created_by_name: String,
    pub updated_by: Option<Uuid>,
    pub updated_by_name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, FromRow)]
pub struct WikiPageUpdate {
    pub id: Uuid,
    pub page_id: Uuid,
    pub client_id: String,
    pub base_version: i64,
    pub version: i64,
    pub patch: serde_json::Value,
    pub content_markdown: String,
    pub created_by: Option<Uuid>,
    pub created_by_name: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, FromRow)]
pub struct WikiPresence {
    pub page_id: Uuid,
    pub client_id: String,
    pub user_id: Option<Uuid>,
    pub display_name: String,
    pub cursor_anchor: Option<i32>,
    pub cursor_head: Option<i32>,
    pub last_seen_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TextPatch {
    pub from: usize,
    pub to: usize,
    pub insert: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateWikiSpaceRequest {
    pub workspace_id: Uuid,
    pub name: String,
    pub slug: String,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateWikiPageRequest {
    pub workspace_id: Uuid,
    pub space_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub title: String,
    pub slug: String,
    pub content_markdown: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ApplyWikiUpdateRequest {
    pub client_id: String,
    pub base_version: i64,
    pub patch: TextPatch,
}

#[derive(Debug, Deserialize)]
pub struct WikiPresenceRequest {
    pub client_id: String,
    pub cursor_anchor: Option<i32>,
    pub cursor_head: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct WikiSyncQuery {
    pub after_version: Option<i64>,
}

#[derive(Debug, Serialize)]
pub struct WikiSyncSnapshot {
    pub page: WikiPage,
    pub updates: Vec<WikiPageUpdate>,
    pub collaborators: Vec<WikiPresence>,
}
